Assembler simplification (pending state removal)

gzanitti opened the pull request:
This PR simplifies the operation of the assembler, mainly by removing the need to maintain a pending state and to return the assembler to its original state in certain situations (which caused issue #124).

The main changes include
- The creation of two structs, PendingMacro and PendingLabel, which store information about the macros and labels used before they were defined, as well as their relative position in the code.
- Some push operations have been changed to take an option<usize> as parameter, and if this is not None, to act more like an insert than a push (useful for example to insert the calls to the pending macros).

Also, issue #124 has been fixed and the corresponding test added.

All tests should pass successfully. However, I would like to test the new assembler a bit more by writing some code (e.g. following @lightclient 's  recommendation and writing an ERC-20 implementation using ETK) before merging it.

In case you don't find it necessary, it can be merged right now.

gzanitti commented:
@lightclient @SamWilsn the assembler rewrite is ready. If anyone has time during the week to review the code, I'll be happy to address your comments :)

P.S: I'm tagging you two because you're the two people I've been talking to about the assembler rewrite. If you don't have time, feel free to ignore this message.

SamWilsn commented:
@gzanitti we're the correct people to tag for sure! I'm swamped for the rest of the week, but I start my vacation next week so I should finally have some time for ETK :tada:

gzanitti commented:
> @gzanitti we're the correct people to tag for sure! I'm swamped for the rest of the week, but I start my vacation next week so I should finally have some time for ETK 🎉

That's great! :). Anyway, today I was playing around with an ERC20 implementation I was working on (basically, moving different parts of the code to macros, adding includes/imports) and I found a bug. Which I thought I had fixed a few hours ago, but it turns out that the new change breaks some "push label"s haha. 

I'll try to solve it in the next few days, but I'm starting to think it might need a bit of a major rewrite. I'll keep you posted.

SamWilsn commented:
That might be the best approach. Keep the pest parser and scrap the rest.

On Wed, Sept 27, 2023, 18:39 Gastón Zanitti ***@***.***>
wrote:

> @gzanitti <https://github.com/gzanitti> we're the correct people to tag
> for sure! I'm swamped for the rest of the week, but I start my vacation
> next week so I should finally have some time for ETK 🎉
>
> That's great! :). Anyway, today I was playing around with an ERC20
> implementation I was working on (basically, moving different parts of the
> code to macros, adding includes/imports) and I found a bug. Which I thought
> I had fixed a few hours ago, but it turns out that the new change breaks
> some "push label"s haha.
>
> I'll try to solve it in the next few days, but I'm starting to think it
> might need a bit of a major rewrite. I'll keep you posted.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/quilt/etk/pull/133#issuecomment-1738199780>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANU4EQOXIUOWPKZQGSYCE5DX4STJXANCNFSM6AAAAAA46VVE2I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gzanitti commented:
> That might be the best approach. Keep the pest parser and scrap the rest.

Excellent. I tried not to do that from the very beginning, because I was trying to preserve as much of the original code as possible (I didn't want to be the outsider who comes to break everything). You're going to say it wasn't necessary, but it's good to have your approval to make a deeper change haha.

I have some ideas I'd like to try that I think will help simplify the whole compiler. 

I'll keep you posted. Thanks!

gzanitti commented:
Small update:

I'm almost done with the new code. The truth is: There are no big changes, just some reorganizations.

Roughly, these are the major changes:

- After the code is parsed, I take care of resolving each include and import, so that the code is as "expanded" as possible.
    - Biggest gain: This allows me to create a new assembler to handle the includes, and makes handling scopes easier. 
    - Biggest loss: After it comes out of the parser, I have to preprocess the code. I tried to do it all in one pass, but the code got too messy and I preferred to prioritize readability (at least for now, we can optimize it in the future). 
    - Important fact: This allowed me to discover that there are two tests where the expected values were wrong.

- I also make sure to save all macros before starting to assemble. Having all the macros predefined saves a lot of headaches, since calling a macro before declaring it means expanding part of the code in the future. Again, maybe it's not as efficient as it could be, but I prefer clarity over performance for now. Everything can be optimized in a future PR.
- Having all the macros already defined also helped me to simplify the assembler I had already defined to remove the pending state at the beginning of this PR, so there is no need to do inserts anymore, all new operations are placed at the end of the list, making the code much simpler and more understandable.

I still need to fix a few things. Most importantly, I would like to add some more tests that make use of the dynamic push of labels (%push(label)) to make sure everything is working properly. Also, the handling and resolution of paths in each include/import is just a "hacky patch". And there is a test that fails when the push label is an expression. Nothing that is likely to be a major headache, but I'd like to have it sorted out before you guys take a look at the code.


Anyway, I'm really enjoying working on this project, so I hope you like my ideas.

P.S.: I'm not a Rust expert, my code might not be "rusty" enough, so I'm open to any suggestions :).

P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)

gzanitti commented:
I think the code is ready :rocket: 
However, I'll be adding a few more tests over the next  days to be sure everything works perfectly.

Thanks!

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -221,21 +211,20 @@ pub struct Assembler {
     /// Labels, in `pending`, that have been referred to (ex. with push) but
     /// have not been declared with an `AbstractOp::Label`.
-    undeclared_labels: HashSet<String>,
+    undefined_labels: Vec<PendingLabel>,
```
It's been a while, so bear with me, but how do we want to use "undeclared" vs "undefined"?

You're probably more familiar with the code than I am at this point, but IIRC, undefined labels were labels that had a `some_label:` in the source code, but didn't have a fixed position yet because of a `%push` macro. Undeclared labels were forward references to a `some_label:` on a later line.

gzanitti replied:
In this case, it's only necessary to keep track of the labels that were pushed before they were declared. I agree with you that the use of the word _undefined_ is not quite correct, so I made the modification so that the field is called _undeclared_labels_. I also corrected the documentation to better understand the meaning of each struct field.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,56 +233,102 @@ impl Assembler {
-                _ => unreachable!(),
-            };
+    /// Create a new `Assembler` with a known length.
+    pub fn new_internal(actual_len: usize) -> Self {
```
If this is an internal function, it should be `pub(crate)`.

---

The Rust naming convention is to use something like `with_len` or `with_length` for a constructor that takes a length.

gzanitti replied:
I was wrong about the two tests I thought were wrong. I fixed both tests and uploaded the assembler modification, so this function is no longer needed.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,56 +233,102 @@ impl Assembler {
-        if !self.ready.is_empty() {
-            panic!("not all assembled bytecode has been taken");
+    /// Inspect the macros in a series of [`RawOp`] and declare them in the assembler.
+    pub fn inspect_macros<I, O>(&mut self, nodes: I) -> Result<(), Error>
```
From a quick read, this looks like an implementation detail? Are you ever going to want to call `push_all` without inspecting the macros?

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,56 +233,102 @@ impl Assembler {
     }
 
     /// Collect any assembled instructions that are ready to be output.
     pub fn take(&mut self) -> Vec<u8> {
```
Did you consider changing the API to just an `assemble` function? I don't think we need the `push_all`/`take` phases if we're only considering the whole program in one chunk.

gzanitti replied:
I fully agree with this. I made the change from push_all to assemble and now the whole process is done through a single call. Even the creation of a new assembler to resolve scopes is now done internally, which makes much more sense.

Thanks!

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,56 +233,102 @@ impl Assembler {
+                    .concretize((&self.declared_labels, &self.declared_macros).into())
+                {
+                    Ok(cop) => cop.assemble(&mut output),
+                    Err(ops::Error::ContextIncomplete { source }) => match source {
```
Is it possible to un-nest the matches? Maybe something like:

```rust
Err(ops::Error::ContextIncomplete { source: UnknownLabel { .. } } => { /* ... */ },
Err(ops::Error::ContextIncomplete { source: UnknownMacro { name, .. } } => { /* ... */ },
Err(ops::Error::ContextIncomplete { source: UndefinedVariable { name, .. } } => { /* ... */ },
```

SamWilsn reviewed (commented):
First few comments. Looking reasonable so far! I'm about halfway through `asm.rs`.

gzanitti commented:
> First few comments. Looking reasonable so far! I'm about halfway through `asm.rs`.

Thanks! Take your time and don't hesitate to ask questions or correct anything you think can be improved. 
Especially if something can be written in a more "rusty" way.

gzanitti commented:
In addition to issue 124, mentioned at the beginning, this PR also fixes issue #108  (both tests mentioned in the issue were added to the test suite).

SamWilsn commented on etk-asm/src/ingest.rs:
```diff
@@ -403,70 +283,87 @@ where
+                        std::fs::read_to_string::<&PathBuf>(&hex_path).with_context(|_| {
+                            error::Io {
+                                message: "reading hex include",
+                                path: <PathBuf as Into<PathBuf>>::into(hex_path.clone()).to_owned(),
```
This seems a bit crazy. Can't just do:

```suggestion
                                path: hex_path,
```

or maybe:

```suggestion
                                path: &hex_path,
```

?

gzanitti replied:
This is rare. This was, after quite a while, the only way I could get the code to work. But I just removed it and it just works :open_mouth:

SamWilsn replied:
snafu sometimes amazes me, and sometimes confounds me.

SamWilsn commented on etk-asm/src/ingest.rs:
```diff
@@ -403,70 +283,87 @@ where
                         .map_err(|e| Box::new(e) as Box<dyn std::error::Error>)
                         .context(error::InvalidHex {
-                            path: partial.path().to_owned(),
+                            path: <PathBuf as Into<PathBuf>>::into(hex_path).to_owned(),
```
Yeah, there's something fishy about the way these conversions are happening.

SamWilsn commented on etk-asm/src/ingest.rs:
```diff
@@ -403,70 +283,87 @@ where
+    fn resolve_and_ingest<P: AsRef<Path>>(
+        &mut self,
+        program: &mut Program,
+        path: P,
```
Generally in Rust, if you immediately clone a reference inside a function, you may as well take the owned type as a parameter. In this case, `PathBuf`.

That lets the caller decide whether to clone or not. If the caller no longer needs the variable, it can transfer ownership (saving the potentially expensive clone.)

gzanitti replied:
I simplified the code a bit so that push_path gets a reference and made some changes to bring the clone to the one place where it was indispensable. Sorry, while I've made a lot of progress in this time, there are still some "common practices" of Rust that escape me.

SamWilsn replied:
I only mentioned it because of your postscript. It's such a tiny thing that I wouldn't've mentioned it otherwise.

gzanitti replied:
Don't worry, these comments are really helpful. There are many things you don't learn in books/tutorials.

SamWilsn commented on etk-asm/src/ingest.rs:
```diff
@@ -403,70 +283,87 @@ where
+    {
+        let path = path.into();
+        let source = program.push_path(path.clone())?;
+        let code = read_to_string::<&PathBuf>(&source).with_context(|_| error::Io {
```
Is the turbofish (`::<>`) necessary here?

gzanitti replied:
To be honest, the compiler asked me to add it so that the code can compile.  But I just removed it as you said and now there is no error. Possibly related to my first implementation. I'm sorry

SamWilsn replied:
Don't be sorry, that's what reviews are for!

SamWilsn commented on etk-asm/src/parse/error.rs:
```diff
@@ -60,6 +62,31 @@ pub enum ParseError {
     },
+
+    /// An argument provided to a macro was of the wrong type.
+    #[snafu(display("File {} does not exist", path.to_string_lossy()))]
```
```suggestion
    #[snafu(display("File `{}` does not exist", path.to_string_lossy()))]
```

SamWilsn commented on etk-asm/src/parse/error.rs:
```diff
@@ -60,6 +62,31 @@ pub enum ParseError {
+    /// An argument provided to a macro was of the wrong type.
+    #[snafu(display("File {} does not exist", path.to_string_lossy()))]
+    #[non_exhaustive]
+    FileNotFound {
```
What returns this variant?

gzanitti replied:
I don't understand this comment, what do you mean?

SamWilsn replied:
If I search through your branch, I only find this definition. I don't see the variant `FileNotFound` being used anywhere. It's entirely possible I'm just missing where it's used, but I wanted to double check before merging.

SamWilsn commented on etk-asm/src/parse/error.rs:
```diff
@@ -60,6 +62,31 @@ pub enum ParseError {
+    /// An included fail failed to parse as hexadecimal.
+    #[snafu(display("included file `{}` is invalid hex: {}", path.to_string_lossy(), source))]
+    #[non_exhaustive]
+    InvalidHex {
```
Similarly, what returns this variant?

SamWilsn reviewed (commented):
Another partial review. Sorry!

gzanitti commented on etk-asm/src/ingest.rs:
```diff
@@ -211,135 +180,48 @@ impl Root {
-        ensure!(self.sources.len() <= 255, error::RecursionLimit);
+    fn push_path(&mut self, path: PathBuf) -> Result<PathBuf, Error> {
+        ensure!(self.depth <= 255, error::RecursionLimit);
+        self.depth += 1;
```
Depth is left over from my first implementation of "Program" when I was only keeping track of the latest source. I have just removed it

gzanitti commented:
> Another partial review. Sorry!

Thank you for your comments @SamWilsn. I just hope I'm not giving you too many headaches haha.

P.S: You can take your time on the review, in the meantime I'm finishing PR #135  and, if necessary, I can spend some time on solving some other items on the list of issues in the meantime.

SamWilsn commented:
> P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)

I missed this part! I'd absolutely love that! There are so many little things I've wanted to build with ETK, and I had basically given up on having the time/motivation to do it. Extremely happy to have someone as excited as you. A website (maybe with an ETK-compiled-to-JavaScript playground :innocent:) is a great idea!

gzanitti commented:
> > P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)
> 
> I missed this part! I'd absolutely love that! There are so many little things I've wanted to build with ETK, and I had basically given up on having the time/motivation to do it. Extremely happy to have someone as excited as you. A website (maybe with an ETK-compiled-to-JavaScript playground 😇) is a great idea!

Great, glad to be able to help, and to discuss and develop all the ideas you have in mind. I'll prepare a prototype in the next few weeks, including the playground you mention, and I'll write to you to see what you think :smile:

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
+                self.ready.clear();
+                v
+            }
+            Err(_) => Vec::new(),
```
This looks like it discards errors from `concretize_ops`.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
-    /// Returns the number of bytes that can be collected with [`Assembler::take`].
-    pub fn push_all<I, O>(&mut self, ops: I) -> Result<usize, Error>
+    /// Returns the code of the assembled program.
+    pub fn assemble<I, O>(&mut self, ops: I) -> Result<Vec<u8>, Error>
```
I think at this point we can just use a slice instead of an iterator.

```suggestion
    pub fn assemble<O>(&mut self, ops: &[O]) -> Result<Vec<u8>, Error>
```

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
+    pub fn assemble<I, O>(&mut self, ops: I) -> Result<Vec<u8>, Error>
     where
-        I: IntoIterator<Item = O>,
+        I: IntoIterator<Item = O> + Clone,
```
```suggestion
```

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
+        Ok(())
+    }
+
+    /// Collect any assembled instructions that are ready to be output.
+    fn take(&mut self) -> Vec<u8> {
+        let output = self.concretize_ops();
+        match output {
+            Ok(v) => {
+                self.ready.clear();
+                v
+            }
+            Err(_) => Vec::new(),
+        }
+    }
```
Let's just move this directly into `assemble`.

```suggestion
```

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
-        Ok(self.ready.len())
+        self.finish()?;
+
+        Ok(self.take())
```
Moving from `take`:

```suggestion
        let output = self.concretize_ops()?;
        self.ready.clear();
        Ok(output)
```

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -244,107 +231,127 @@ impl Assembler {
-                                .collect::<Vec<String>>();
-                            error::UndeclaredLabels { labels: missing }.fail()
+    /// Inspect the macros in a series of [`RawOp`] and declare them in the assembler.
+    fn inspect_macros<I, O>(&mut self, nodes: &I) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = O> + Clone,
+        O: Into<RawOp>,
+    {
+        for op in nodes.clone() {
```
Ah, yeah, I see now. You're doing two passes, so you need to clone the iterator. I suggested changing `assemble` to just take a slice, so no more cloneable iterator.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -353,39 +360,59 @@ impl Assembler {
         let rop = rop.into();
 
-        self.declare_content(&rop)?;
+        self.declare_label(&rop)?;
 
-        // Expand instruction macros immediately. We do this here because it's the same process
-        // regardless if we `push_read` or `push_pending` -- in fact, `expand_macro` pushes each op
-        // individually which calls the correct unchecked push.
+        // Expand instruction macros immediately.
         if let RawOp::Op(AbstractOp::Macro(ref m)) = rop {
             self.expand_macro(&m.name, &m.parameters)?;
-            return Ok(self.ready.len());
+            return Ok(self.concrete_len);
         }
 
-        self.push_unchecked(rop)?;
-        Ok(self.ready.len())
+        self.push_rawop(rop)?;
+        Ok(self.concrete_len)
     }
```
Is it possible to combine `push` with `push_rawop`? Like move `self.expand_macro(&m.name, &m.parameters)?;` (line 373) to line 424?

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -394,26 +421,15 @@ impl Assembler {
-                }
-                Ok(())
-            }
+            RawOp::Op(AbstractOp::Macro(_)) => Ok(()),
```
If I understand this correctly, hitting a macro invocation inside this function is a bug, right? We should either merge `push` and `push_rawop`, or put `unreachable!()` here instead of `Ok(())`.

gzanitti replied:
Not really. While they have all been processed at this point, they are not removed from the original code. So this is simply to ignore them.

gzanitti replied:
Yes, it's not exactly a bug, but since the macro had already been processed before, it was simply ignored. I will combine both pushes to simplify the code.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -430,203 +446,48 @@ impl Assembler {
-                        self.pending_len = rop.size();
-                        self.pending.push_back(rop);
+                    Err(ops::Error::ContextIncomplete {
+                        source: UnknownLabel { label: _label, .. },
```
```suggestion
                        source: UnknownLabel { .. },
```

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -430,203 +446,48 @@ impl Assembler {
+                        match op.size() {
+                            Some(size) => self.concrete_len += size,
+                            None => {
+                                self.concrete_len += 2;
```
If `op` is unsized, does it make sense to add two here? If I'm assembling something like:

```
%push(faraway)
%repeat_jumpdest(511)
faraway:
```

The push would have to be three bytes long.

gzanitti replied:
Yes, exactly, but in this instance it works more like a placeholder. Then, when the operations are actually concretised and the final position of the label is known, it is corrected. That's why also the dynamic jump flag is set to true. 

In fact, there is a test to prove that this works: 
https://github.com/gzanitti/etk/blob/12c609c081bc96812b067b12a45192c1f700e3c9/etk-asm/tests/asm.rs#L297

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -430,203 +446,48 @@ impl Assembler {
+                        source: UnknownLabel { label: _label, .. },
+                    }) => {
+                        let mut dynamic_push = false;
+                        match op.size() {
```
Why do you use `size` to determine if this is a variable push? A more accurate way would be something like:

```rust
if let AbstractOp::Push(_) = op { ... }
```

gzanitti replied:
My idea was to make it as general as possible. I was not only interested in pushes, but in any other operation that won't have a defined size in the future. But rethinking it, you'll probably have to modify other parts of the code if that happens, so the idea of using if let AbstractOp::Push(_) makes a lot of sense. Thanks!

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -353,58 +353,43 @@ impl Assembler {
+                for ul in self.undeclared_labels.iter() {
+                    if (ul.label == label) & ul.dynamic_push {
+                        // Compensation in case label was dynamically pushed.
+                        let mut tmp =
+                            ((self.concrete_len as f32 - ul.position as f32) / 256.0).floor();
```
Why use floating point for this? It seems equivalent to `(self.concrete_len - ul.position) / 256`. Integer division in Rust always floors.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -353,58 +353,43 @@ impl Assembler {
+                        let mut tmp =
+                            ((self.concrete_len as f32 - ul.position as f32) / 256.0).floor();
+
+                        // Size already accounted for %push(label) was 2.
+                        if tmp > 1.0 {
+                            tmp -= 1.0;
+                        }
+                        dst = cmp::max(tmp as usize, dst);
```
Something smells off about this, but I can't quite put my finger on it. Let me try a few things, and I'll be back.

gzanitti replied:
Sure! Let me know if you have any questions or if I can help you in any way. Basically that part of the code takes care of compensating the value of the labels. If a dynamic push was used to define the label in question, in that instance its position is already known, so it "stretches" the value to reach the final position. As all intermediate macros were already expanded because they are known beforehand, there should be no problem in setting their value correctly.

But if you find any interesting edge case, feel free to let me know and I can add the test to be sure it works.

gzanitti replied:
Oh, I see, I didn't realise that labels could be defined that way. Excellent catch. I'll fix the code as quickly as possible.

SamWilsn replied:
My thought is that these two programs must assemble to the same thing:

<table>
<tr>
<th>
</th>
<th>Constant Push</th>
<th>Variable Push</th>
</tr>
<tr>
<th>Input</th>
<td>

```
push2 label + 254                   
label:
pc
```

</td>
<td>

```
%push(label + 254)                   
label:
pc
```

</td>
</tr>

<tr>
<th>Output</th>
<td>

```
61010158
```

</td>
<td>

```
61010058
```

</td>
</tr>
</table>

I assumed the constant push was correct, but regardless, they should be the same.

gzanitti replied:
Oh, I see. I'll check it out. Thanks!

gzanitti replied:
Just to be sure. This should be a valid program too, right?

```
%push(label1 + label2 + 254)
label:
pc
label2:
pc
```

gzanitti replied:
Nothing out of the ordinary; I was just working on the previous case and noticed that this one wasn't included in my solution, and I thought it should be. I wanted to double-check. I'll figure out a solution that covers both, but it might mean changing some parts you've already looked at. Sorry about that.

gzanitti replied:
Solved. There were many intermediate steps, so I simplified the code a bit and renamed things for better understanding.

gzanitti commented on etk-asm/src/asm.rs:
```diff
@@ -430,202 +311,137 @@ impl Assembler {
+                                .collect();
+                            if !labels.is_empty() {
+                                if let Ok(val) = exp {
+                                    let extra = val.to_usize().unwrap() / 256;
```
Hi Sam. I improved a little the documentation of the function and added some comments so that you can better understand what's happening in each step. I hope it's clearer now :)

SamWilsn commented:
Sorry it's taken me so long to re-review! I believe I found another case that passed before this pull request and is failing now.

Thanks again for being so persistent on this. I _really_ appreciate the help getting ETK into a good place <3

gzanitti commented:
> Sorry it's taken me so long to re-review! I believe I found another case that passed before this pull request and is failing now.
> 
> Thanks again for being so persistent on this. I _really_ appreciate the help getting ETK into a good place <3

Sorry, it's also my fault for having to change several parts of the code in the middle of the process and make you check everything again.  As I always say, thanks for your patience and for letting me collaborate :)

Regarding the code, I solved the case you mention and also added some more tests. The novelty is that now all the pushes are checked and updated taking into account the previous updates (acum variable) (declared_labels is now a BTreeMap because it matters the order when adding the accumulated size to the labels).

gzanitti commented:
Hey @SamWilsn, I needed a HashMap sorted by insert and ended up adding IndexMap to the project. However, it is only used in the assembler and could be replaced by a struct with a HashMap and a Vector if you don't want to add more dependencies. 

Let me know what you think.

SamWilsn reviewed (commented):
Uggh, I found another test case that fails with this approach. I'll push it up in a second. I'm so sorry x.x

SamWilsn commented:
I'm totally happy with the `IndexMap`.

SamWilsn commented:
Pushed the new test. Summary is:

<table>
<tr><td></td><th>Variable + Fixed</th><th>Variable + Variable</th></tr>
<tr>
<th>Source</th>

<td>

```
%push(lbl1 - lbl2)
push2 lbl1 + lbl2
pc # repeat 126 times.
lbl1:
lbl2:
```

</td>
<td>

```
%push(lbl1 - lbl2)
%push(lbl1 + lbl2)
pc # repeat 126 times.
lbl1:
lbl2:
```

</td>
</tr>
<tr>
<th>Outcome</th>
<td>

`6000610106...`

</td>
<td>

`6000610108...`

</td>
</tr>
</table>

gzanitti commented:
Hey @SamWilsn , how are you?
First of all, sorry for the delay. I've been having some personal issues and couldn't dedicate as much time as I would have liked to ETK. Also, my country is slowly going into the abyss, which doesn't help.

Don't worry about mentioning as many corrections as you deem necessary. After all, this code belongs more to you than to me and I can make all the changes you consider necessary until you are satisfied :).

Regarding the bug you reported, I'm sorry I didn't see it before, I understand that with the last change nothing like this should happen again.

I look forward to your comments!

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -430,202 +343,163 @@ impl Assembler {
+                    .eval_with_context((&self.declared_labels, &self.declared_macros).into());
+
+                if let Ok(val) = exp {
+                    let imm_size = (val.to_f64().unwrap().log2() / 8.0).ceil() as usize;
```
Can we use [`BigInt::bits`](https://docs.rs/num-bigint/latest/num_bigint/struct.BigInt.html#method.bits) and [integer ceiling division](https://stackoverflow.com/a/2745086) here instead of converting to a float? I'm not sure if this'll give the correct answer for really big numbers.

SamWilsn commented on etk-asm/src/asm.rs:
```diff
@@ -430,202 +343,163 @@ impl Assembler {
+                    let imm_size = (val.to_f64().unwrap().log2() / 8.0).ceil() as usize;
+
+                    if imm_size > 1 {
+                        for (_label, label_value) in self.declared_labels.iter_mut() {
```
Would [`IndexMap::values_mut`](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.values_mut) be more appropriate?

SamWilsn reviewed (commented):
Just these last two items!
